include/nil/crypto3/zk/snark/arithmetization/plonk/detail/lookup_table_definition.hpp

```diff
@@ -0,0 +1,166 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//---------------------------------------------------------------------------//
+
+#ifndef CRYPTO3_ZK_PLONK_DETAIL_LOOKUP_TABLE_DEFINITION_HPP
+#define CRYPTO3_ZK_PLONK_DETAIL_LOOKUP_TABLE_DEFINITION_HPP
+
+#include <string>
+
+#include <nil/crypto3/zk/snark/arithmetization/plonk/constraint_system.hpp>
+
+namespace nil {
+    namespace crypto3 {
+        namespace zk {
+            namespace snark {
+                namespace detail {                    
+                    // Interf-ace for lookup table definitions.
+                    template<typename FieldType>
+                    class lookup_subtable_definition{
+                    public:
+                        std::vector<std::size_t> column_indices;
+                        std::size_t begin;
+                        std::size_t end;
+                    };
+
+                    template<typename FieldType>
+                    class lookup_table_definition{
+                    protected:
+                        std::vector<std::vector<typename FieldType::value_type>> _table;
+                    public:
+                        std::string table_name;
+                        std::map<std::string, lookup_subtable_definition<FieldType>> subtables;
+
+                        lookup_table_definition(const std::string table_name){
+                            this->table_name = table_name;
+                        }                        
+
+                        virtual void generate() = 0;
+
+                        const std::vector<std::vector<typename FieldType::value_type>> &get_table(){
+                            if(_table.size() == 0){
+                                generate();
+                            }
+                            return _table;
+                        }
+                        virtual ~lookup_table_definition() {};
+                    };
+
+                    template<typename FieldType>
+                    class filled_lookup_table_definition:public lookup_table_definition<FieldType>{
+                    public:
+                        filled_lookup_table_definition(std::shared_ptr<lookup_table_definition<FieldType>> &other):lookup_table_definition<FieldType>(other->table_name){

```

Comment:
Why do we need this constructor from a shared_ptr? It looks strange, better not have it, let the caller send a simple reference as in usual copy constructor.